oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/hierarchy/HierarchyService.java

```diff
@@ -60,84 +67,149 @@ private Map<String, List<String>> getHierarchyDefinition() {
-     * relations should be defined in `config/hierarchy-definition.yml`. Generally, if instance A and instance B could detect
-     * each other, then build the hierarchy in one side is enough.
+     * relations should be defined in `config/hierarchy-definition.yml`.
      *
-     * @param instanceName        the name of the source instance (self)
-     * @param serviceName         the name of the source service (self)
-     * @param serviceLayer        the layer of the source service
-     * @param relatedInstanceName the name of the detected related instance
-     * @param relatedServiceName  the name of the detected related service
-     * @param relatedServiceLayer the layer of the detected related service
+     * @param upperInstanceName        the name of the upper instance
+     * @param upperServiceName         the name of the upper service
+     * @param upperServiceLayer        the layer of the upper service
+     * @param lowerInstanceName the name of the lower related instance
+     * @param lowerServiceName  the name of the lower related service
+     * @param lowerServiceLayer the layer of the lower related service
      */
-    public void toInstanceHierarchyRelation(String instanceName,
-                                            String serviceName,
-                                            Layer serviceLayer,
-                                            String relatedInstanceName,
-                                            String relatedServiceName,
-                                            Layer relatedServiceLayer) {
+    public void toInstanceHierarchyRelation(String upperInstanceName,
+                                            String upperServiceName,
+                                            Layer upperServiceLayer,
+                                            String lowerInstanceName,
+                                            String lowerServiceName,
+                                            Layer lowerServiceLayer) {
         if (!this.isEnableHierarchy) {
             return;
         }
-        if (!checkHierarchyDefinition(serviceLayer, relatedServiceLayer)) {
+        Map<String, HierarchyDefinitionService.MatchingRule> lowerLayers = getHierarchyDefinition().get(upperServiceLayer.name());
+        if (lowerLayers == null || !lowerLayers.containsKey(lowerServiceLayer.name())) {
+            log.error("upperServiceLayer " + upperServiceLayer.name() + " or lowerServiceLayer " + lowerServiceLayer.name()
+                          + " is not defined in hierarchy-definition.yml.");
             return;
         }
+
+        buildInstanceHierarchyRelation(upperInstanceName, upperServiceName, upperServiceLayer, lowerInstanceName,
+                                           lowerServiceName, lowerServiceLayer);
+    }
+
+    public void startAutoMatchingServiceHierarchy() {
+        if (!this.isEnableHierarchy) {
+            return;
+        }
```

Comment:
Let's move this check to the provider `CoreModuleProvider` so we don't bother to initialize everything that is never used.